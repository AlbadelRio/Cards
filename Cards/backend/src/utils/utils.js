function DataChanges(body, userData, userFound) {
  switch (body.type) {
    case 'cardsAdquired':
      userData.cardsAdquired.push(body.cardId);
      userFound.save();
      break;

    case 'difficultCards':
      userData.difficultCards.push(body.cardId);
      userFound.save();
      break;
    case 'stats': {
      const existingDate = userData.stats.find((data) => data.date === body.date);
      if (!existingDate) {
        userData.stats.push(body);
        userFound.save();
      } else {
        existingDate.stats.update(
          {
            $push: { timePlayed: body.time, percentages: body.percentage },
            $sort: { score: -1 }
          }

        );
      } }
      break;
    default:
      break;
  }
}

module.exports = { DataChanges };
