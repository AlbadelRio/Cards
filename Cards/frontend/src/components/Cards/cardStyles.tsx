import { StyleSheet } from 'react-native';

export default StyleSheet.create({
  background: {
    width: '100%',
    height: '100%',
    backgroundColor: '#d5d5d5',
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center'
  },
  scrollBackground: {
    width: '75%',
    height: '100%',
    backgroundColor: '#ffffff',
    padding: 10
  },
  h1: {
    fontSize: 35,
    fontWeight: 'bold',
    color: '#272727'
  },
  tinyLogo: {
    width: 45,
    height: 45
  },
  tinyDelete: {
    width: 15,
    height: 15
  },
  delete: {
    width: 45,
    height: 45,
    backgroundColor: '#5EBBB0',
    borderRadius: 3,
    marginRight: 10,
    alignItems: 'center',
    justifyContent: 'center',
    elevation: 7
  },
  items: {
    width: 400,
    height: 100,
    flex: 1,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'flex-start'
  },
  info: {
    justifyContent: 'flex-start',
    alignItems: 'center',
    flexDirection: 'column',
    marginRight: 15,
    color: '#272727'
  },
  cards: {
    flexDirection: 'row',
    alignItems: 'center',
    marginRight: 10
  },
  text: {
    fontWeight: 'bold',
    fontSize: 15,
    color: '#272727'
  },
  add: {
    width: 80,
    height: 40,
    backgroundColor: '#5EBBB0',
    borderRadius: 3,
    elevation: 7,
    fontWeight: 'bold',
    fontSize: 18,
    textAlign: 'center',
    padding: 8,
    color: '#272727',
    marginTop: 20

  },
  button: {
    flex: 1,
    alignItems: 'flex-end'
  }
});
