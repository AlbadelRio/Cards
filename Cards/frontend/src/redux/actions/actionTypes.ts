export default {
  AUTH_LOGIN: 'AUTH_LOGIN',
  AUTH_LOGOUT: 'AUTH_LOGOUT',
  AUTH_LOGIN_ERROR: 'AUTH_LOGIN_ERROR',
  ERROR_GENERIC: 'ERROR_GENERIC',
  REFRESH_TOKEN: 'REFRESH_TOKEN',
  LOAD_PACKCARDS: 'LOAD_PACKCARDS',
  GETRANDOM_PACKCARD: 'GETRANDOM_PACKCARD',
  ADD_SUBSCRIPTOR_PACKCARD: 'ADD_SUBSCRIPTOR_PACKCARD'

};
