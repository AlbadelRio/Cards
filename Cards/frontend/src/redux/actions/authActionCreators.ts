import axios from 'axios';

import actionTypes from './actionTypes';

type User= {
  email:string,
  password:string
}

export function login(user:User) {
  console.log(user);
  return async (dispatch:any) => {
    console.log(dispatch);
    try {
      const { data } = await axios.post('http://192.168.0.103:5000/login', user);
      console.log('login:');
      console.log(data);
      return dispatch({
        type: actionTypes.AUTH_LOGIN,
        user: data
      });
    } catch (error:any) {
      console.log(error);
      if (error?.response?.status === 401) {
        return dispatch({
          type: actionTypes.AUTH_LOGIN_ERROR,
          message: error.message
        });
      }
      return dispatch({
        type: actionTypes.ERROR_GENERIC,
        message: error.message
      });
    }
  };
}
export function userRefreshToken(refreshToken:string) {
  return async (dispatch:any) => {
    try {
      console.log('aaaa');
      const { data } = await axios.post('http://192.168.0.103:5000/refreshToken', { refreshToken });
      console.log('data userRefreshToken');

      console.log(data);
      return dispatch({
        type: actionTypes.REFRESH_TOKEN,
        token: data
      });
    } catch (error:any) {
      if (error?.response?.status === 401) {
        return dispatch({
          type: actionTypes.AUTH_LOGIN_ERROR,
          message: error.message
        });
      }
      return dispatch({
        type: actionTypes.ERROR_GENERIC,
        message: error.message
      });
    }
  };
}

export function logOut() {
  return {
    type: actionTypes.AUTH_LOGOUT
  };
}
