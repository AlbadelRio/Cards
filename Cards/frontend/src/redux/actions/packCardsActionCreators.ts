import axios from 'axios';

import actionTypes from './actionTypes';

export function loadPackcards() {
  return async (dispatch:any) => {
    const { data } = await axios.get('http://192.168.0.24:5000/api/packCards');
    return dispatch({
      type: actionTypes.LOAD_PACKARDS,
      packCards: data
    });
  };
}

export function getRandomPackCardBySubject(subject:string) {
  return async (dispatch:any) => {
    const { data } = await axios.get(`http://192.168.0.24:5000/api/packCards?subject=${subject}`);
    return dispatch({
      type: actionTypes.GETRANDOM_PACKCARD,
      randomPackCard: data
    });
  };
}
