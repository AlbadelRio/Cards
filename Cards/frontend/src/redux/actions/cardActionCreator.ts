import axios from 'axios';
import { ENV_VAR } from '@env';
import { userRefreshToken } from './authActionCreators';
import actionTypes from './actionTypes';

export function createCard(token:any, refreshtoken:any, activePackCard:string, body:any) {
  return async (dispatch:any) => {
    if (token) {
      const { data } = await axios.post(ENV_VAR.concat(`/api/cards/${activePackCard}`), body, {
        headers: { Authorization: `Bearer ${token}` }
      });
      dispatch({
        type: actionTypes.CREATE_CARD,
        newCard: data

      });
    }
    return userRefreshToken(refreshtoken);
  };
}

export function updateCard(token:any, refreshtoken:any, activeCard:string, body:any) {
  return async (dispatch:any) => {
    if (token) {
      const { data } = await axios.put(ENV_VAR.concat(`/api/cards/${activeCard}`), body, {
        headers: { Authorization: `Bearer ${token}` }
      });
      dispatch({
        type: actionTypes.UPDATE_CARD,
        newCard: data

      });
    }
    return userRefreshToken(refreshtoken);
  };
}
