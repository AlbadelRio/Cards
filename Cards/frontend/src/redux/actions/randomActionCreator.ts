import axios from 'axios';
import { ENV_VAR } from '@env';
import { userRefreshToken } from './authActionCreators';
import actionTypes from './actionTypes';

export default function getRandomPackCardBySubject(token:any, refreshtoken:any, subject:string) {
  return async (dispatch:any) => {
    if (token) {
      const { data } = await axios.get(ENV_VAR.concat(`/api/packCards?subject=${subject}`), {
        headers: { Authorization: `Bearer ${token}` }
      });
      return dispatch({
        type: actionTypes.GETRANDOM_PACKCARD,
        randomPackCard: data
      });
    }
    return userRefreshToken(refreshtoken);
  };
}
