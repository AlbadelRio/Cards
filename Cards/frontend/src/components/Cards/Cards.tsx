/* eslint-disable no-underscore-dangle */
import React, { useEffect, useState } from 'react';
import {
  View, Text, Pressable, Image, ScrollView
} from 'react-native';
import LinearGradient from 'react-native-linear-gradient';
import { useDispatch, useSelector } from 'react-redux';

import { deletePackCard, loadPackcards } from '../../redux/actions/packCardsActionCreators';
import styles from './cardStyles';

export default function Cards({ navigation }:any) {
  const { token, refreshToken } = useSelector((store:any) => store.tokensReducer);
  const userId = useSelector((store:any) => store.auth.user.user._id);
  const packCards = useSelector((store:any) => store.packardsReducer);
  const subscriptedPack = useSelector((store:any) => store.subscribeToPackCardReducer);

  const dispatch = useDispatch();
  const [allUserPackCards, setAllUserPackCards] = useState([]);

  useEffect(() => {
    dispatch(loadPackcards(token, refreshToken));
  }, [packCards.length, packCards.packCards, subscriptedPack]);

  useEffect(() => {
    const subscripted = packCards?.reduce((acc: any, packCard: any) => {
      if (packCard?.subscriptors
        .some(({ userId: subscriptorUserId }: any) => subscriptorUserId === userId)) {
        return [...acc, packCard];
      }
      return acc;
    }, []);

    const owned = packCards?.filter((value:any) => value?.user === userId);
    setAllUserPackCards(owned.concat(subscripted));
  }, [packCards]);

  function deleteHandler(pack:any) {
    dispatch(deletePackCard(
      token,
      refreshToken,
      pack._id
    ));
  }
  return (
    <View style={styles.background}>
      <View style={styles.scrollBackground}>
        <Text style={styles.h1}>MY CARDS</Text>
        <View>
          <ScrollView>
            <Text>
              {allUserPackCards.length}
              {' '}
              PACKCARDS
            </Text>
            {allUserPackCards.map((pack:any, index:any) => (
              <Pressable
                testID={`update${index}`}
                key={pack._id}
                style={styles.items}
                onPress={() => {
                  navigation.navigate('UpdateForm', {
                    pack
                  });
                }}
              >
                <LinearGradient
                  start={{ x: 1, y: 0 }}
                  end={{ x: 0, y: 1 }}
                  colors={['#5EBBB0', '#5D5FEF']}
                  style={styles.linearGradient}
                >
                  <Pressable
                    testID={`delete${index}`}
                    onPress={() => { deleteHandler(pack); }}
                  >
                    <View>
                      <Image
                        style={styles.tinyDelete}
                        source={{
                          uri: 'https://img.icons8.com/plumpy/24/000000/waste.png'
                        }}
                      />
                    </View>
                  </Pressable>
                </LinearGradient>
                <View style={styles.cards}>
                  <View style={styles.info}>
                    <Text style={styles.text}>{pack.title}</Text>
                    <Text>{pack.subject}</Text>
                  </View>
                  <View>
                    <Text>
                      {pack?.packCards?.length}
                      {' '}
                      CARDS
                    </Text>
                  </View>
                </View>
                <Pressable
                  testID={`carroussel${index}`}
                  onPress={() => {
                    navigation.navigate('Carroussel', {
                      userId,
                      pack
                    });
                  }}
                >
                  <View>
                    <Image
                      style={styles.tinyLogo}
                      source={{
                        uri: 'https://img.icons8.com/material-rounded/96/000000/play--v1.png'
                      }}
                    />
                  </View>
                </Pressable>
              </Pressable>
            ))}

            <View style={styles.button}>
              <LinearGradient
                start={{ x: 1, y: 0 }}
                end={{ x: 0, y: 1 }}
                colors={['#5EBBB0', '#5D5FEF']}
                style={styles.linearGradientButton}
              >
                <Pressable
                  testID="add"
                  onPress={() => {
                    navigation.navigate('CreationForm', {
                      userId,
                      token,
                      refreshToken
                    });
                  }}
                >
                  <Text style={styles.add}>ADD</Text>

                </Pressable>
              </LinearGradient>
            </View>
          </ScrollView>
        </View>
      </View>
    </View>
  );
}
