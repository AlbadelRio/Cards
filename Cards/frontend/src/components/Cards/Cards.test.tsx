/* eslint-disable no-underscore-dangle */
import React from 'react';
import { render, fireEvent } from '../../utils/test.utils';
import { deletePackCard } from '../../redux/actions/packCardsActionCreators';
import Cards from './Cards';

jest.mock('../../redux/actions/packCardsActionCreators', () => ({
  ...jest.requireActual('../../redux/actions/packCardsActionCreators'),
  deletePackCard: jest.fn()
}));

describe('Given a Cards component', () => {
  describe('when it is rendered', () => {
    let screen:any;
    let navigation:any;
    beforeEach(() => {
      const initialState = {
        auth: {
          user: {
            user: {
              _id: '1'
            }
          }
        },
        packardsReducer: [
          {
            _id: '0',
            user: '3',
            subscriptors: [{ userId: '1' }, { userId: '3' }]
          },
          {
            _id: '14',
            user: '1',
            subscriptors: [{ userId: '2' }]
          }
        ]
      };

      navigation = { navigate: jest.fn() };
      screen = render(<Cards navigation={navigation} />, initialState);
    });
    test('Then should match the snapshot', () => {
      expect(screen).toMatchSnapshot();
    });
    test('Should call navigation.navigate', () => {
      const pressable = screen.getByTestId('update0');
      fireEvent.press(pressable);
      expect(navigation.navigate).toHaveBeenCalled();
    });
    test('Should call navigation.navigate', () => {
      const pressable = screen.getByTestId('carroussel0');
      fireEvent.press(pressable);
      expect(navigation.navigate).toHaveBeenCalled();
    });
    test('should call navigation.navigate', () => {
      const pressable = screen.getByTestId('add');
      fireEvent.press(pressable);
      expect(navigation.navigate).toHaveBeenCalled();
    });
    test('should call deleteHandler', () => {
      (deletePackCard as jest.Mock).mockReturnValueOnce({ type: '' });
      const pressable = screen.getByTestId('delete0');
      fireEvent.press(pressable);

      expect(deletePackCard).toHaveBeenCalled();
    });
  });
});
