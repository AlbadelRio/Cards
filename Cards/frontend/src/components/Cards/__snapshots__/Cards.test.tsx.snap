// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a Cards component when it is rendered Then should match the snapshot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#d5d5d5",
      "flex": 1,
      "height": "100%",
      "justifyContent": "center",
      "width": "100%",
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#ffffff",
        "height": "100%",
        "padding": 10,
        "width": "75%",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#272727",
          "fontSize": 35,
          "fontWeight": "bold",
        }
      }
    >
      MY CARDS
    </Text>
    <View>
      <RCTScrollView>
        <View>
          <Text>
            2
             
            PACKCARDS
          </Text>
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "row",
                "height": 100,
                "justifyContent": "space-between",
                "width": 300,
              }
            }
            testID="update0"
          >
            <BVLinearGradient
              colors={
                Array [
                  4284398512,
                  4284309487,
                ]
              }
              endPoint={
                Object {
                  "x": 0,
                  "y": 1,
                }
              }
              locations={null}
              startPoint={
                Object {
                  "x": 1,
                  "y": 0,
                }
              }
              style={
                Object {
                  "alignItems": "center",
                  "borderRadius": 3,
                  "elevation": 7,
                  "height": 45,
                  "justifyContent": "center",
                  "width": 45,
                }
              }
            >
              <View
                accessible={true}
                collapsable={false}
                focusable={true}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                testID="delete0"
              >
                <View>
                  <Image
                    source={
                      Object {
                        "uri": "https://img.icons8.com/plumpy/24/000000/waste.png",
                      }
                    }
                    style={
                      Object {
                        "height": 15,
                        "width": 15,
                      }
                    }
                  />
                </View>
              </View>
            </BVLinearGradient>
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                  "width": 200,
                }
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "color": "#272727",
                    "flexDirection": "column",
                    "justifyContent": "flex-start",
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "color": "#272727",
                      "fontSize": 15,
                      "fontWeight": "bold",
                    }
                  }
                />
                <Text />
              </View>
              <View>
                <Text>
                   
                  CARDS
                </Text>
              </View>
            </View>
            <View
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              testID="carroussel0"
            >
              <View>
                <Image
                  source={
                    Object {
                      "uri": "https://img.icons8.com/material-rounded/96/000000/play--v1.png",
                    }
                  }
                  style={
                    Object {
                      "height": 45,
                      "width": 45,
                    }
                  }
                />
              </View>
            </View>
          </View>
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "row",
                "height": 100,
                "justifyContent": "space-between",
                "width": 300,
              }
            }
            testID="update1"
          >
            <BVLinearGradient
              colors={
                Array [
                  4284398512,
                  4284309487,
                ]
              }
              endPoint={
                Object {
                  "x": 0,
                  "y": 1,
                }
              }
              locations={null}
              startPoint={
                Object {
                  "x": 1,
                  "y": 0,
                }
              }
              style={
                Object {
                  "alignItems": "center",
                  "borderRadius": 3,
                  "elevation": 7,
                  "height": 45,
                  "justifyContent": "center",
                  "width": 45,
                }
              }
            >
              <View
                accessible={true}
                collapsable={false}
                focusable={true}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                testID="delete1"
              >
                <View>
                  <Image
                    source={
                      Object {
                        "uri": "https://img.icons8.com/plumpy/24/000000/waste.png",
                      }
                    }
                    style={
                      Object {
                        "height": 15,
                        "width": 15,
                      }
                    }
                  />
                </View>
              </View>
            </BVLinearGradient>
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                  "width": 200,
                }
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "color": "#272727",
                    "flexDirection": "column",
                    "justifyContent": "flex-start",
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "color": "#272727",
                      "fontSize": 15,
                      "fontWeight": "bold",
                    }
                  }
                />
                <Text />
              </View>
              <View>
                <Text>
                   
                  CARDS
                </Text>
              </View>
            </View>
            <View
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              testID="carroussel1"
            >
              <View>
                <Image
                  source={
                    Object {
                      "uri": "https://img.icons8.com/material-rounded/96/000000/play--v1.png",
                    }
                  }
                  style={
                    Object {
                      "height": 45,
                      "width": 45,
                    }
                  }
                />
              </View>
            </View>
          </View>
          <View
            style={
              Object {
                "alignItems": "flex-end",
                "flex": 1,
              }
            }
          >
            <BVLinearGradient
              colors={
                Array [
                  4284398512,
                  4284309487,
                ]
              }
              endPoint={
                Object {
                  "x": 0,
                  "y": 1,
                }
              }
              locations={null}
              startPoint={
                Object {
                  "x": 1,
                  "y": 0,
                }
              }
              style={
                Object {
                  "borderRadius": 4,
                  "elevation": 7,
                  "height": 50,
                  "marginTop": 20,
                  "width": 90,
                }
              }
            >
              <View
                accessible={true}
                collapsable={false}
                focusable={true}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                testID="add"
              >
                <Text
                  style={
                    Object {
                      "color": "#272727",
                      "fontSize": 18,
                      "fontWeight": "bold",
                      "padding": 12,
                      "textAlign": "center",
                    }
                  }
                >
                  ADD
                </Text>
              </View>
            </BVLinearGradient>
          </View>
        </View>
      </RCTScrollView>
    </View>
  </View>
</View>
`;
